[lldb][NFC] Split RegisterContextUnwind::SavedLocationForRegister

jasonmolenda opened the pull request:
RegisterContextUnwind::SavedLocationForRegister is around 450 lines that first find an abstract register location (e.g. "CFA-8") for a register by looking in the UnwindPlans.  Then it evaluates the abstract register location to create a concrete register location (e.g. "stored at address 0x...", "live in register at frame 0"). There are some complicated cases in the first half of the method to handle return address register architectures correctly, in particular.

Looking at the two halves, they're both exactly 226 lines long and there's little involvement between them except for passing an abstract register location along.

(there were some parts in the "abstract register location" code that would set the concrete register location, unnecessarily)

It's also a complex enough method that there are some bits of code that aren't actually doing anything at this point.

This patch adds a RegisterContextUnwind::GetAbstractRegisterLocation method, which does the first half, and has a clearly defined return values.

The code to convert an AbstractRegisterLocation into a ConcreteRegisterLocation remains in SavedLocationForRegister.

It's a bit of a tricky patch to visually inspect, despite it not changing functionality, the reorganizations and rewrites make the diff unreadable.  Nearly all the real changes are in the "find the abstract register location" first half of the method. I think reading the new code in its new form is the easiest way to inspect this PR.  With a defined interface between the two of what is expected, it's pretty easy to look at the code and reason about whether it is written correctly.

(whereas before, that was very difficult, for me at least.)

jasonmolenda committed:
[lldb][NFC] Split RegisterContextUnwind::SavedLocationForRegister

RegisterContextUnwind::SavedLocationForRegister is around 450 lines
that first find an abstract register location (e.g. "CFA-8") for a
register by looking in the UnwindPlans.  Then it evaluates the
abstract register location to create a concrete register location
(e.g. "stored at address 0x...", "live in register at frame 0").
There are some complicated cases in the first half of the method
to handle return address register architectures correctly, in
particular.

Looking at the two halves, they're both exactly 226 lines long and
there's little involvement between them except for passing an
abstract register location along.

(there were some parts in the "abstract register location" code
that would set the concrete register location, unnecessarily)

It's also a complex enough method that there are some bits of code
that aren't actually doing anything at this point.

This patch adds a RegisterContextUnwind::GetAbstractRegisterLocation
method, which does the first half, and has a clearly defined return
values.

The code to convert an AbstractRegisterLocation into a
ConcreteRegisterLocation remains in SavedLocationForRegister.

It's a bit of a tricky patch to visually inspect, despite it not
changing functionality, the reorganizations and rewrites make the
diff unreadable.  Nearly all the real changes are in the
"find the abstract register location" first half of the method.
I think reading the new code in its new form is the easiest way
to inspect this PR.  With a defined interface between the two of
what is expected, it's pretty easy to look at the code and reason
about whether it is written correctly.

(whereas before, that was very difficult, for me at least.)

jasonmolenda committed:
Remove another bit of code in GetAbstractRegisterLocation
that wasn't doing anything.

jasonmolenda committed:
Don't log a message when the ABI says this is a volatile register
-- has an Undefined abstract register location.

jasonmolenda commented:
to be clear: The diff for the first half of the old SavedLocationForRegister is entirely pointless to look at.  Reading RegisterContextUnwind::GetAbstractRegisterLocation in the "new" version of the file is the way to go, and I'm pretty comfortable with everything that I've kept in here, I think the behaviors are easy to reason as correct when I read it.

The diff for the second half -- taking an AbstractRegisterLocation and converting it into a ConcreteRegisterLocation -- involved very few changes and is readable.

I've run it through our testsuite on macOS and it shows no issues, but I look forward to all the CI bots getting their hands on it & seeing that result confirmed.

jasonmolenda committed:
clarify comment.

jasonmolenda committed:
make another pass over all the comments and
rewrite/clarify a bunch of them.  Tiny tweaks to
the code to make it more readable.

jasonmolenda committed:
tiny edit

labath commented on lldb/source/Target/RegisterContextUnwind.cpp:
```diff
@@ -1243,303 +1243,285 @@ bool RegisterContextUnwind::IsTrapHandlerSymbol(
+
+  // Have we already found this register location?
+  if (!m_registers.empty()) {
+    std::map<uint32_t,
+             lldb_private::UnwindLLDB::ConcreteRegisterLocation>::const_iterator
+        iterator;
+    iterator = m_registers.find(regnum.GetAsKind(eRegisterKindLLDB));
```
this is one of the few officially sanctioned uses of auto

```suggestion
    auto iterator = m_registers.find(regnum.GetAsKind(eRegisterKindLLDB));
```

jasonmolenda committed:
Update lldb/source/Target/RegisterContextUnwind.cpp

Co-authored-by: Pavel Labath <pavel@labath.sk>

jasonmolenda committed:
Still undecided about the documentation of GetAbstractRegisterLocation.
I do want it next to the method because I think it makes it clear
(to people editing the method) what the expected behaviors are,
more than people who might be *calling* the method and looking at
the header file.  But it is unusual so we'll see what other people
think.

jasonmolenda committed:
typeo

jasonmolenda committed:
Merge branch 'main' into savedlocationforregister-nfc-cleanup

jasonmolenda committed:
Merge branch 'main' into savedlocationforregister-nfc-cleanup
